pkg/kv/kvnemesis/validator.go

```diff
@@ -587,28 +579,37 @@ func (v *validator) processOp(buffering bool, op Operation) {
 // checkAtomic verifies a set of operations that should be atomic by trying to find
 // a timestamp at which the observed reads and writes of the operations (as executed
 // in the order in which they appear in the arguments) match the MVCC history.
-func (v *validator) checkAtomic(
-	atomicType string, result Result, optOpsTimestamp hlc.Timestamp, ops ...Operation,
-) {
+func (v *validator) checkAtomic(atomicType string, result Result, ops ...Operation) {
 	for _, op := range ops {
 		// NB: we're not really necessarily in a txn, but passing true here means that
 		// we have an atomic unit, which is also the case if we are called here by a
 		// non-transactional Put, for example.
 		v.processOp(isBuffering, op)
 	}
-	txnObservations := v.curOps
-	v.curOps = nil
+	observations := v.curObservations
+	v.curObservations = nil
 
 	if result.Type != ResultType_Error {
-		v.checkCommittedTxn(`committed `+atomicType, txnObservations, optOpsTimestamp)
+		// The timestamp is not optional in this case. Note however that at the time
+		// of writing, checkAtomicCommitted doesn't capitalize on this unconditional
+		// presence yet, and most unit tests don't specify it for reads.
+		execTimestamp := result.OptionalTimestamp
```

Comment:
Should we assert this isn't empty?

Reply:
This comment is unfortunately incorrect, since we still have AmbiguousResultError + when the unambiguous operation had only deletions we won't be able to guess the timestamp. Will update this before merge.

Reply:
Ah nevermind it's correct here. Will add a check.